jl4-lsp/src/LSP/L4/Viz/Ladder.hs

```diff
@@ -30,67 +49,108 @@ isDevMode = False
   , substitution   :: !TC.Substitution  -- might be more futureproof to use TypeCheckResult
   , shouldSimplify :: !Bool             -- ^ whether to simplify the expression
   }
   deriving stock (Show, Generic, Eq)
 
-data VizState =
-  MkVizState
-    { env :: !VizEnv
-    , maxId :: !ID
-    }
-  deriving stock (Show, Generic, Eq)
 
-mkInitialVizState :: VizEnv -> VizState
-mkInitialVizState env =
+data VizState = MkVizState
+  { cfg           :: !VizConfig
+  , maxId         :: !ID
+  , evalAppMakers :: !(Map Int (V.EvalAppRequestParams -> TopDecl Resolved))
+  -- ^ Map from Unique of V.ID to eval-app-directive maker
+  }
+  deriving stock (Generic)
+
+instance Show VizState where
+  show MkVizState{cfg, maxId, evalAppMakers} =
+    "MkVizState { cfg = " <> show cfg <>
+    ", maxId = " <> show maxId <>
+    ", (keys of) evalAppMakers = " <> show (Map.keys evalAppMakers) <> " }"
+
+mkInitialVizState :: VizConfig -> VizState
+mkInitialVizState cfg =
   MkVizState
-    { env = env
+    { cfg = cfg
     , maxId = MkID 0
+    , evalAppMakers = Map.empty
     }
 
 -- Monad ops
 
 -- | 'Internal' helper: This should only be used by other Viz monad ops
-getVizEnv :: Viz VizEnv
-getVizEnv = use #env
+getVizCfg :: Viz VizConfig
+getVizCfg = use #cfg
 
 getVerTxtDocId :: Viz LSP.VersionedTextDocumentIdentifier
 getVerTxtDocId = do
-  env <- getVizEnv
-  pure env.verTxtDocId
+  cfg <- getVizCfg
+  pure cfg.verTxtDocId
 
 getExpandedType :: Type' Resolved -> Viz (Type' Resolved)
 getExpandedType ty = do
-  env <- getVizEnv
-  pure $ TC.applyFinalSubstitution env.substitution env.moduleUri ty
+  cfg <- getVizCfg
+  pure $ TC.applyFinalSubstitution cfg.substitution cfg.moduleUri ty
 
 getFresh :: Viz ID
 getFresh = do
   #maxId <%= \(MkID n) -> MkID (n + 1)
 
 getShouldSimplify :: Viz Bool
 getShouldSimplify = do
-  env <- getVizEnv
-  pure env.shouldSimplify
+  cfg <- getVizCfg
+  pure cfg.shouldSimplify
+
+{- | Assumes that the program is well-scoped and well-typed -}
+prepEvalAppMaker :: V.ID -> Expr Resolved -> Viz ()
+prepEvalAppMaker vid = \ case
+  App appAnno appResolved _ -> do
+    localDecls <- getLocalDecls
+    let maker = \V.EvalAppRequestParams{args} ->
+          Directive emptyAnno $ LazyEval emptyAnno $
+            Where emptyAnno
+              (App appAnno appResolved $ map toBoolExpr args)
+              localDecls
+    #evalAppMakers %= Map.insert vid.id maker
+  _ -> pure ()
+
+getLocalDecls :: Viz [LocalDecl Resolved]
+getLocalDecls = do
+  env <- ask
+  pure env.localDecls
+
+withLocalDecls :: [LocalDecl Resolved] -> Viz a -> Viz a
+withLocalDecls newLocalDecls = local (\env -> env { localDecls = newLocalDecls <> env.localDecls })
+
+-- Viz state helpers
+
+lookupEvalAppMaker :: VizState -> V.ID -> Maybe (V.EvalAppRequestParams -> TopDecl Resolved)
+lookupEvalAppMaker vs vid = Map.lookup vid.id vs.evalAppMakers
+
+getVizConfig :: VizState -> VizConfig
+getVizConfig vs = vs.cfg
```

Comment:
I don't understand why you use this mix of record dot accessors and functions that are defined in terms of them.

Reply:
I wanted to hide the implementational details of VizState (e.g. how VizState is related to VizConfig).

I've pushed a commit with a comment explaining this, and removed the export of the internals of VizState.